web_transport/sdk/impl/http3_server_session.cc

```diff
@@ -0,0 +1,123 @@
+                            supported_versions,
+                            connection,
+                            visitor,
+                            helper,
+                            crypto_config,
+                            compressed_certs_cache),
+      backend_(backend),
+      io_runner_(io_runner),
+      event_runner_(event_runner) {
+  CHECK(io_runner_);
+  CHECK(event_runner_);
+}
+
+::quic::QuicSpdyStream* Http3ServerSession::CreateIncomingStream(
+    ::quic::QuicStreamId id) {
+  if (!ShouldCreateIncomingStream(id)) {
+    return nullptr;
+  }
+  std::unique_ptr<::quic::QuicSpdyStream> stream =
+      std::make_unique<Http3ServerStream>(id, this,
+                                          ::quic::StreamType::BIDIRECTIONAL,
+                                          backend_, io_runner_, event_runner_);
+  ::quic::QuicSpdyStream* stream_ptr = stream.get();
+  ActivateStream(std::move(stream));
+  return stream_ptr;
+}
+
+::quic::QuicSpdyStream* Http3ServerSession::CreateIncomingStream(
+    ::quic::PendingStream* pending) {
+  std::unique_ptr<::quic::QuicSpdyStream> stream =
+      std::make_unique<Http3ServerStream>(pending, this,
+                                          ::quic::StreamType::BIDIRECTIONAL,
+                                          backend_, io_runner_, event_runner_);
+  ::quic::QuicSpdyStream* stream_ptr = stream.get();
+  ActivateStream(std::move(stream));
+  return stream_ptr;
+}
+
+::quic::QuicSpdyStream*
+Http3ServerSession::CreateOutgoingBidirectionalStream() {
+  if (!WillNegotiateWebTransport()) {
+    LOG(ERROR)
+        << "CreateOutgoingBidirectionalStream willout WebTransport support.";
+    return nullptr;
+  }
+  if (!ShouldCreateOutgoingBidirectionalStream()) {
+    return nullptr;
+  }
+
+  ::quic::QuicServerInitiatedSpdyStream* stream =
+      new ::quic::QuicServerInitiatedSpdyStream(
+          GetNextOutgoingBidirectionalStreamId(), this, ::quic::BIDIRECTIONAL);
+  ActivateStream(absl::WrapUnique(stream));
+  return stream;
+}
+
+::quic::QuicSpdyStream*
+Http3ServerSession::CreateOutgoingUnidirectionalStream() {
+  if (!ShouldCreateOutgoingUnidirectionalStream()) {
+    return nullptr;
+  }
+
+  std::unique_ptr<::quic::QuicSpdyStream> stream =
+      std::make_unique<Http3ServerStream>(
+          GetNextOutgoingUnidirectionalStreamId(), this,
+          ::quic::StreamType::WRITE_UNIDIRECTIONAL, backend_, io_runner_,
+          event_runner_);
+  ::quic::QuicSpdyStream* stream_ptr = stream.get();
+  ActivateStream(std::move(stream));
+  return stream_ptr;
+}
+
+std::unique_ptr<::quic::QuicCryptoServerStreamBase>
+Http3ServerSession::CreateQuicCryptoServerStream(
+    const ::quic::QuicCryptoServerConfig* crypto_config,
+    ::quic::QuicCompressedCertsCache* compressed_certs_cache) {
+  return CreateCryptoServerStream(crypto_config, compressed_certs_cache, this,
+                                  stream_helper());
+}
+
+bool Http3ServerSession::ShouldNegotiateWebTransport() {
```

Comment:
Can you explain why these two APIs are placed here, and do you have plan to really implement them?

Reply:
It overrides QuicSpdySession's methods. Returning `true` means certain HTTP session support WebTransport negotiation. Not all HTTP sessions support WebTransport.